src/contract.jl

```diff
@@ -36,3 +40,66 @@ function contract_density_matrix(
   end
   return out
 end
+
+function ITensors.scalar(
+  alg::Algorithm, tn::Union{AbstractITensorNetwork,Vector{ITensor}}; kwargs...
+)
+  return contract(alg, tn; kwargs...)[]
+end
+
+function ITensors.scalar(
+  tn::Union{AbstractITensorNetwork,Vector{ITensor}}; alg="exact", kwargs...
+)
+  return scalar(Algorithm(alg), tn; kwargs...)
+end
+
+function logscalar(
+  tn::Union{AbstractITensorNetwork,Vector{ITensor}}; alg::String="exact", kwargs...
+)
+  return logscalar(Algorithm(alg), tn; kwargs...)
+end
+
+function logscalar(
+  alg::Algorithm"exact", tn::Union{AbstractITensorNetwork,Vector{ITensor}}; kwargs...
```

Comment:
I think it would be better to define a type alias:
```julia
const AnyITensorNetwork = Union{AbstractITensorNetwork,Vector{ITensor}}
```
and then use `::AnyITensorNetwork` here and elsewhere.

Reply:
Though there is some question about when and why we should allow inputs of both `AbstractITensorNetwork` and `Vector{ITensor}`, I was generally trying to be conservative about that and require users to wrap `Vector{ITensor}` into `ITensorNetwork` themselves. Is there a particular reason for allowing that here?

Reply:
Yeah that is a good question. I think here I was taking the stance that lots of times those functions end up passing down to `contract(tns::Vector{ITensor})` so if the user wants `logscalar` on a `Vector{ITensor}` with `alg = "exact"` then they don't have to wrap it up as it will just get unwrapped.

I guess the main point is, depending on the algorithm, certain `scalar()` calls will unwrap to a `Vector{ITensor}` and pass down to something like `contract` whilst others will want to wrap up to an `ITensorNetwork` and pass to new code like `logscalar(alg::Algorithm"BP", ...)` so I was thinking of a generalized interface for calling either

Reply:
That's good to try and keep in mind but probably we should assess that more systematically and decide which kinds of functions should be able to accept `Vector{ITensor}` vs. `AbstractITensorNetwork` in a future PR. I could imagine changing many constraints from `AbstractITensorNetwork` to `AnyITensorNetwork` to make them more general when it makes sense, but let's think about that for a future PR.

Reply:
Okay I will leave as is for now.

Reply:
I misunderstood your plan of action for this, my vote would be to restrict these functions to `tn::ITensorNetwork` for now and think about extending to other collections of ITensors like `::Vector{ITensor}` in future work.

Reply:
I see, I thought you meant to just leave as is. I will change the functions in this script to only apply to `tn::AbstractITensorNetwork`. I will remove `contract_density_matrix(contract_list::Vector{ITensor}, ...)` and just put it in the `test_belief_propagation.jl` file as that is the only place it is being used (to define an update routine for the message tensors based on `approx_itensornetwork`)

Reply:
Makes sense, thanks. We probably need to refactor/redesign the approximate contraction backend anyway since it is a bit of a mess.